src/protocols/http/hpr_roaming_protocol.erl

```diff
@@ -0,0 +1,541 @@
+        erlang:integer_to_binary(PacketTime),
+        DestURL,
+        erlang:atom_to_binary(FlowType)
+    ],
+    Token0 = lists:join(?TOKEN_SEP, Parts),
+    Token1 = erlang:iolist_to_binary(Token0),
+    hpr_roaming_utils:binary_to_hexstring(Token1).
+
+-spec parse_uplink_token(token()) ->
+    {ok, transaction_id(), region(), non_neg_integer(), dest_url(), flow_type()} | {error, any()}.
+parse_uplink_token(<<"0x", Token/binary>>) ->
+    parse_uplink_token(Token);
+parse_uplink_token(Token) ->
+    Bin = hpr_roaming_utils:hex_to_binary(Token),
+    case binary:split(Bin, ?TOKEN_SEP, [global]) of
+        [TransactionIDBin, RegionBin, PacketTimeBin, DestURLBin, FlowTypeBin] ->
+            TransactionID = erlang:binary_to_integer(TransactionIDBin),
+            Region = erlang:binary_to_atom(RegionBin),
+            PacketTime = erlang:binary_to_integer(PacketTimeBin),
+            FlowType = erlang:binary_to_existing_atom(FlowTypeBin),
+            {ok, TransactionID, Region, PacketTime, DestURLBin, FlowType};
+        _ ->
+            {error, malformed_token}
+    end.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec select_best(list(packet())) -> packet().
+select_best(Copies) ->
+    [Best | _] = lists:sort(
+        fun(#packet{packet_up = PacketUpA}, #packet{packet_up = PacketUpB}) ->
+            RSSIA = hpr_packet_up:rssi(PacketUpA),
+            RSSIB = hpr_packet_up:rssi(PacketUpB),
+            RSSIA > RSSIB
+        end,
+        Copies
+    ),
+    Best.
+
+-spec gw_info(packet()) -> map().
+gw_info(#packet{packet_up = PacketUp}) ->
+    PubKeyBin = hpr_packet_up:gateway(PacketUp),
+    Region = hpr_packet_up:region(PacketUp),
+
+    SNR = hpr_packet_up:snr(PacketUp),
+    RSSI = hpr_packet_up:rssi(PacketUp),
+
+    GW = #{
+        'ID' => hpr_roaming_utils:binary_to_hexstring(hpr_gwmp_worker:pubkeybin_to_mac(PubKeyBin)),
+        'RFRegion' => Region,
+        'RSSI' => erlang:trunc(RSSI),
+        'SNR' => SNR,
+        'DLAllowed' => true
+    },
+    GW.
+
+-spec encode_deveui(non_neg_integer()) -> binary().
+encode_deveui(Num) ->
+    hpr_roaming_utils:hexstring(Num, 16).
+
+-spec encode_devaddr(non_neg_integer()) -> binary().
+encode_devaddr(Num) ->
+    hpr_roaming_utils:hexstring(Num, 8).
+
+-spec window(non_neg_integer(), 'undefined' | non_neg_integer(), atom()) ->
+    packet_router_pb:window_v1_pb().
+window(TS, Freq, DataRate) ->
+    #window_v1_pb{timestamp = TS, frequency = Freq, datarate = DataRate}.
+
+-spec new_downlink(
+    Payload :: binary(),
+    Timestamp :: non_neg_integer(),
+    Frequency :: atom() | number(),
+    DataRate :: atom() | integer(),
+    Rx2 :: packet_router_pb:window_v1_pb() | undefined
+) -> downlink_packet().
+new_downlink(Payload, Timestamp, Frequency, DataRate, Rx2) ->
+    #packet_router_packet_down_v1_pb{
```

Comment:
If we can, let's keep the records out of the protocol code.

And we need to make sure our Frequencies are mapped correctly.

Http -> Mhz
Proto -> Hz

Reply:
@michaeldjeffrey 
I have reviewed and corrected the Frequency mapping.

Reply:
Let's try and keep all our interactions with the `packet_up` and `packet_down` records in the modules `hpr_packet_up` and `hpr_packet_down`.